# Day 3 - Logistic Regression & Classification Metrics

## Introduction

Logistic Regression is a supervised machine learning algorithm used for classification problems. It predicts the probability that a sample belongs to a specific class and then classifies it into one of the possible categories.

## Objective

The goal of this notebook is to train a Logistic Regression classifier, generate predictions, evaluate its performance using classification metrics, and interpret the results.

## What is Logistic Regression?

Logistic Regression is a supervised learning algorithm used for classification tasks.

Unlike Linear Regression, which predicts continuous numerical values, Logistic Regression predicts class labels such as 0 or 1.

The model calculates a weighted sum of the input features and then applies the sigmoid function to convert the result into a probability between 0 and 1.

## Dataset

I used the Pima Indians Diabetes Dataset.

The target variable is **Outcome**.

- Outcome = 0 → No Diabetes
- Outcome = 1 → Diabetes

Since the target contains categories, this is a classification problem.

In [2]:
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (confusion_matrix,classification_report,roc_auc_score)

In [7]:
df = pd.read_csv("diabetes.csv")
df.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1
3,1,89,66,23,94,28.1,0.167,21,0
4,0,137,40,35,168,43.1,2.288,33,1


I checked the DataSet inofrmation,shape,columns names,data type, and missing values befor training :

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB


In [9]:
df.shape

(768, 9)

In [10]:
df.columns

Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='str')

In [11]:
df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

there is no missing values in the dataset so its ready for training 

## Hands - on lab : Building a Classifier 



##  Select Features and Target

The target variable is **Outcome**.

The selected features are all health-related measurements except the target column.

These features will be used to predict whether a patient has diabetes.


In [12]:
x = df.drop("Outcome",axis=1)
y = df["Outcome"]

In [13]:
x.head()

,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age
0,6,148,72,35,0,33.6,0.627,50
1,1,85,66,29,0,26.6,0.351,31
2,8,183,64,0,0,23.3,0.672,32
3,1,89,66,23,94,28.1,0.167,21
4,0,137,40,35,168,43.1,2.288,33


In [14]:
y.head()

0    1
1    0
2    1
3    0
4    1
Name: Outcome, dtype: int64

### Observation

X contains all input features that describe each patient.

y contains the Outcome values (0 = No Diabetes, 1 = Diabetes), which the model will learn to predict.

## Train/Test Split

The dataset is divided into training and testing sets.

80% of the data is used for training, while 20% is reserved for testing.

The parameter `stratify=y` ensures that both the training and testing sets keep approximately the same proportion of diabetic and non-diabetic patients.

In [16]:
x_train, x_test, y_train, y_test = train_test_split(
    x,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [17]:
print("X train shape:", x_train.shape)
print("X test shape:", x_test.shape)
print("y train shape:", y_train.shape)
print("y test shape:", y_test.shape)

X train shape: (614, 8)
X test shape: (154, 8)
y train shape: (614,)
y test shape: (154,)


## Step 1: Train a LogisticRegression model on a provided classification dataset (e.g. churn).

A Logistic Regression model is created and trained using the training dataset.

The model learns the relationship between the input features and the target variable to classify whether a patient has diabetes.

In [18]:
model = LogisticRegression(max_iter=1000)
model.fit(x_train,y_train)
print("model trained successfully")

model trained successfully


### Observation

The Logistic Regression model was successfully trained using the training dataset.

During training, the model learned the relationship between the health-related features and the Outcome variable.

The trained model is now ready to classify new patients as diabetic or non-diabetic.

## Step 2: Generate predictions and produce the confusion matrix.

After training the model,predictions are generated for the test dataset

In [ ]:
pre = model.predict(x_test)
pre[:10] 


array([1, 0, 0, 0, 0, 0, 0, 1, 0, 1])

0 = NO diabets 
1 = Diabets 

In [20]:
comp =pd.DataFrame({
    "actual" :y_test.values,
    "predected":pre
})
comp.head(10)

,actual,predected
0,0,1
1,0,0
2,0,0
3,1,0
4,0,0
5,0,0
6,1,0
7,1,1
8,0,0
9,0,1


### Observation

The model generated predictions for the test dataset.

Each prediction is classified as either 0 (No Diabetes) or 1 (Diabetes).

Some predictions match the actual values, while others may differ, which is expected in real-world classification problems.

## Confusion Matrix

The confusion matrix shows how many predictions were correct and incorrect for each class.

In [21]:
cm = confusion_matrix(y_test,pre)
print(cm)

[[82 18]
 [26 28]]


[[TN FP]
 [FN TP]]

In [22]:
cm_df = pd.DataFrame(
    cm,
    index=["Actual 0", "Actual 1"],
    columns=["Predicted 0", "Predicted 1"]
)

cm_df

,Predicted 0,Predicted 1
Actual 0,82,18
Actual 1,26,28


### Observation

The confusion matrix shows that the model correctly classified **82 non-diabetic patients (True Negatives)** and **28 diabetic patients (True Positives)**.

However, the model incorrectly predicted **18 non-diabetic patients as diabetic (False Positives)** and missed **26 diabetic patients by predicting them as non-diabetic (False Negatives)**.

Overall, the model correctly classified more samples than it misclassified, but reducing the number of false negatives would improve its performance for diabetes prediction.

 ## Step 3: Compute precision, recall, and F1 with classification_report, and interpret each.

In [23]:
print(classification_report(y_test,pre))

              precision    recall  f1-score   support

           0       0.76      0.82      0.79       100
           1       0.61      0.52      0.56        54

    accuracy                           0.71       154
   macro avg       0.68      0.67      0.67       154
weighted avg       0.71      0.71      0.71       154



### Observation

The model works better for class 0 than class 1.

It correctly identifies most non-diabetic patients, but it misses some diabetic patients because the recall for class 1 is lower.

The recall for class 1 still needs improvement.

## STEP 4 : Decide whether precision or recall matters more for this specific problem and justify it.

For diabetes prediction, recall is more important than precision.
A higher recall helps the model detect more patients who actually have diabetes
Missing a diabetic patient is more serious than predicting a healthy patient as diabetic.

## Step 5: Compute the AUC-ROC and document what it says about the model.

In [24]:
prob = model.predict_proba(x_test)[:,1]
auc = roc_auc_score(y_test,prob)
print("Auc , Roc ",auc )

Auc , Roc  0.8229629629629629


The AUC-ROC score is 0.823, which is a good score.

This means the model can usually separate diabetic and non-diabetic patients.